editor/src/utils/react-performance.ts

```diff
@@ -209,141 +209,188 @@ function failSafeMemoEqualityFunction(componentDisplayName: string, severity: 's
 
-  const newValueStack: Array<any> = [...valueStackSoFar, possibleNewValueInner]
+  if (oldValue === possibleNewValue) return oldValue
 
   if (
-    oldValueInner &&
-    possibleNewValueInner &&
-    typeof oldValueInner == 'object' &&
-    typeof possibleNewValueInner == 'object'
+    oldValue &&
+    possibleNewValue &&
+    typeof oldValue == 'object' &&
+    typeof possibleNewValue == 'object'
   ) {
-    const arrA = isArray(oldValueInner)
-    const arrB = isArray(possibleNewValueInner)
-
-    // reusable vars for all the performance
-    let i: number = 0
-    let length: number = 0
-    let key: string = ''
+    if (oldValue.constructor !== possibleNewValue.constructor) return possibleNewValue
 
-    if (arrA && arrB) {
-      length = possibleNewValueInner.length
+    var length, i, keys
+    if (Array.isArray(oldValue)) {
+      let newArrayToReturn: any[] = []
+      let canSaveOldArray = true
 
-      var newArrayToReturn: any[] = []
-      var canSaveOldArray = true
-      if (length != oldValueInner.length) {
+      length = possibleNewValue.length
+      if (length != oldValue.length) {
         canSaveOldArray = false
       }
       for (i = length; i-- !== 0; ) {
         // try to recurse into the array item here and save it if possible
         newArrayToReturn[i] = keepDeepReferenceEqualityInner(
-          oldValueInner[i],
-          possibleNewValueInner[i],
+          oldValue[i],
+          possibleNewValue[i],
           stackSizeInner + 1,
-          newValueStack,
+          valueStackSoFar,
         )
-        if (oldValueInner[i] !== newArrayToReturn[i]) {
+        if (oldValue[i] !== newArrayToReturn[i]) {
           canSaveOldArray = false
         }
       }
       if (canSaveOldArray) {
-        return oldValueInner
+        return oldValue
       } else {
         return newArrayToReturn
       }
     }
 
-    if (arrA != arrB) return possibleNewValueInner
-
-    var dateA = oldValueInner instanceof Date,
-      dateB = possibleNewValueInner instanceof Date
-    if (dateA != dateB) return possibleNewValueInner
-    if (dateA && dateB)
-      return oldValueInner.getTime() == possibleNewValueInner.getTime()
-        ? oldValueInner
-        : possibleNewValueInner
-
-    var regexpA = oldValueInner instanceof RegExp,
-      regexpB = possibleNewValueInner instanceof RegExp
-    if (regexpA != regexpB) return possibleNewValueInner
-    if (regexpA && regexpB)
-      return oldValueInner.toString() == possibleNewValueInner.toString()
-        ? oldValueInner
-        : possibleNewValueInner
-
-    // for objects we do a deep recursion
-    var keys = keyList(possibleNewValueInner)
+    if (oldValue instanceof Map && possibleNewValue instanceof Map) {
+      let canSaveOldMap = true
+      let newMapToReturn: Map<any, any> = new Map()
+
+      if (oldValue.size !== possibleNewValue.size) {
+        canSaveOldMap = false
+      }
+      for (i of possibleNewValue.entries()) {
```

Comment:
For the sake of clarity, given how often `i` is usually used for an index in a for loop.
```suggestion
      for (entry of possibleNewValue.entries()) {
```

Reply:
+1

Reply:
fixed